bigchaindb/core.py

```diff
@@ -553,41 +556,54 @@ def vote(self, block, previous_block_id, decision, invalid_reason=None):
 
         return vote_signed
 
-    def write_vote(self, block, vote, block_number):
+    def write_vote(self, block, vote):
         """Write the vote to the database."""
 
         # First, make sure this block doesn't contain a vote from this node
         if self.has_previous_vote(block):
             return None
 
-        # We need to *not* override the existing block_number, if any
-        # FIXME: MIGHT HAVE RACE CONDITIONS WITH THE OTHER NODES IN THE FEDERATION
-        if 'block_number' not in vote:
-            vote['block_number'] = block_number  # maybe this should be in the signed part...or better yet, removed..
-
         r.table('votes') \
             .insert(vote) \
             .run(self.conn)
 
     def get_last_voted_block(self):
         """Returns the last block that this node voted on."""
 
-        last_voted = r.table('votes') \
-            .filter(r.row['node_pubkey'] == self.me) \
-            .order_by(r.desc('block_number')) \
-            .limit(1) \
-            .run(self.conn)
-
-        # return last vote if last vote exists else return Genesis block
-        if not last_voted:
+        try:
+            # get the latest value for the vote timestamp (over all votes)
+            max_timestamp = r.table('votes') \
+                .filter(r.row['node_pubkey'] == self.me) \
+                .concat_map(lambda x: [x['vote']['timestamp']]) \
+                .max() \
+                .run(self.conn)
+
+            last_voted = list(r.table('votes') \
```

Comment:
There should be less results for a given `timestamp` then for a `node_pubkey`. It is probably more efficient to first filter by `timestamp` and then by `node_pubkey`

Also `r.row['vote']['timestamp']` and `lambda x: x['vote']['timestamp']` seem to be equivalent, maybe we should stick to one notation.